doc/neovim: better document the wrappers (#373805)
NixOS/nixpkgs: doc/languages-frameworks/neovim.section.md

# Neovim {#neovim}

Install `neovim-unwrapped` to get a barebone neovim to configure imperatively.
This is the closest to what you encounter on other distributions.

`neovim` is a wrapper around neovim with some extra configuration to for
instance set the various language providers like python.
The wrapper can be further configured to include your favorite plugins and
configurations for a reproducible neovim across machines.
See the next section for more details.

## Custom configuration {#neovim-custom-configuration}

There are two wrappers available to provide additionnal configuration around the vanilla package `pkgs.neovim-unwrapped`:
1. `wrapNeovim`: the historical one you should use
2. `wrapNeovimUnstable` intended to replace the former. It has more features but
   the interface is not stable yet.

You can configure the former via:

```nix
neovim.override {
  configure = {
    customRC = ''
      # here your custom configuration goes!
    '';
    packages.myVimPackage = with pkgs.vimPlugins; {

>       # see examples below how to use custom packages

Reviewer: ```suggestion
      # See examples below on how to use custom packages.
```

      start = [ ];

>       # If a Vim plugin has a dependency that is not explicitly listed in
>       # opt that dependency will always be added to start to avoid confusion.

Reviewer: ```suggestion
      # If a Vim plugin has a dependency that is not explicitly listed in
      # `opt`, that dependency will always be added to `start` to avoid confusion.
```

For instance, `sqlite-lua` needs `g:sqlite_clib_path` to be set to work. Nixpkgs historically patched these in the plugins with several drawbacks: harder maintainance and making upstream work harder. Per convention, these mandatory bits of configuration are bookmarked in nixpkgs in `passthru.initLua`. Enabling `autoconfigure` adds automatically the snippets required for the plugins to work.
- `autowrapRuntimeDeps`: append to PATH runtime dependencies of your plugins. For instance `rest.nvim` requires `curl` to work. Enabling `autowrapRuntimeDeps` adds it to the PATH visible by your neovim wrapper (but not your global PATH).
  neovim wrapper.
- `luaRcContent`: extra lua code to add to the generated `init.lua`
- `neovimRcContent`: extra vimL code sourced by the generated init.lua
- `wrapperArgs`: extra arguments forwarded to the `makeWrapper` call
- `plugins`: a list of plugins to add to the wrapper.

```
wrapNeovimUnstable {
  autoconfigure = true;
  autowrapRuntimeDeps = true;
  plugins = with vimPlugins; [
    {
      plugin = vim-obsession;
      config = ''
        map <Leader>$ <Cmd>Obsession<CR>
      '';
    }
    hex-nvim
  ];
}
```

> You can explore t configuration in the `nix repl` to discover these options and

Reviewer: ```suggestion
You can explore the configuration in the `nix repl` to discover these options and
```

}
```

You can use the new unstable wrapper but the interface may change:
- `autoconfigure`: certain plugins need a custom configuration to work with nix.

> For instance, `sqlite-lua` needs `g:sqlite_clib_path` to be set to work. Nixpkgs historically patched these in the plugins with several drawbacks: harder maintainance and making upstream work harder. Per convention, these mandatory bits of configuration are bookmarked in nixpkgs in `passthru.initLua`. Enabling `autoconfigure` adds automatically the snippets required for the plugins to work.

Reviewer: ```suggestion
For instance, `sqlite-lua` needs `g:sqlite_clib_path` to be set to work. Nixpkgs historically patched these in the plugins with several drawbacks: harder maintenance and making upstream work harder. Per convention, these mandatory bits of configuration are bookmarked in nixpkgs in `passthru.initLua`. Enabling `autoconfigure` automatically adds the snippets required for the plugins to work.
```

> - `autowrapRuntimeDeps`: append to PATH runtime dependencies of your plugins. For instance `rest.nvim` requires `curl` to work. Enabling `autowrapRuntimeDeps` adds it to the PATH visible by your neovim wrapper (but not your global PATH).

Reviewer: ```suggestion
- `autowrapRuntimeDeps`: Appends plugin's runtime dependencies to `PATH`. For instance, `rest.nvim` requires `curl` to work. Enabling `autowrapRuntimeDeps` adds it to the `PATH` visible by your Neovim wrapper (but not your global `PATH`).
```

>   neovim wrapper.

Reviewer: ? Leftover or heading?

>   neovim wrapper.

Author: ```suggestion
```

> - `luaRcContent`: extra lua code to add to the generated `init.lua`
> - `neovimRcContent`: extra vimL code sourced by the generated init.lua
> - `wrapperArgs`: extra arguments forwarded to the `makeWrapper` call
> - `wrapRc`: nix not being able to write in your $HOME, nixpkgs loads the
>   generated neovim configuration via its  `-u` argument, i.e. : `-u /nix/store/...generatedInit.lua`. This has sideeffects like preventing neovim to read your config in `$XDG_CONFIG_HOME` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in neovim). Disable it if you want to generate your own wrapper. You can still reuse while reusing the logic of the nixpkgs wrapper and access the generated config via `neovim.passthru.initRc`.
> - `plugins`: a list of plugins to add to the wrapper.

Reviewer: ```suggestion
- `luaRcContent`: Extra lua code to add to the generated `init.lua`.
- `neovimRcContent`: Extra vimL code sourced by the generated `init.lua`.
- `wrapperArgs`: Extra arguments forwarded to the `makeWrapper` call.
- `wrapRc`: Nix, not being able to write in your `$HOME`, loads the
  generated Neovim configuration via its  `-u` argument, i.e. : `-u /nix/store/...generatedInit.lua`. This has side effects like preventing Neovim from reading your config in `$XDG_CONFIG_HOME` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in Neovim). Disable it if you want to generate your own wrapper. You can still reuse while reusing the logic of the nixpkgs wrapper and access the generated config via `neovim.passthru.initRc`.
- `plugins`: A list of plugins to add to the wrapper.
```

```
wrapNeovimUnstable {
  autoconfigure = true;
  autowrapRuntimeDeps = true;
  luaRcContent = ''
    vim.o.sessionoptions = 'buffers,curdir,help,tabpages,winsize,winpos,localoptions'
    vim.g.mapleader = ' '
    vim.g.maplocalleader = ' '
    vim.opt.smoothscroll = true
    vim.opt.colorcolumn = { 100 }
    vim.opt.termguicolors = true
  '';
  # plugins accepts a list of either plugins or { plugin = ...; config = ..vimscript.. };
  plugins = with vimPlugins; [
    {
      plugin = vim-obsession;
      config = ''
        map <Leader>$ <Cmd>Obsession<CR>
      '';
    }
    (nvim-treesitter.withPlugins (p: [ p.nix p.python ]))
    hex-nvim
  ];
}
```

> You can explore the configuration in the `nix repl` to discover these options and

Reviewer: ```suggestion
You can explore the configuration with`nix repl` to discover these options and
```
